### DISCUSSION TITLE
About the use of CacheWriter after it has been removed

### DISCUSSION THREAD

**Discussion by WeiMingzhi** [ben-manes/caffeine#686] [Q&A] [answered]
Hello,
I found during the upgrade from 2.x to 3.x that CacheWriter has been removed, 
and the 2.x version has annotated '`Scheduled for removal in version 3.0.0. Consider instead using Map compute methods for extending manual write and remove operations, and using Caffeine.evictionListener(RemovalListener) for extending removals due to eviction.`'. 
I know how to use the listener when removing, and I don't see how to implement the extension of the put operation. 
Please excuse my lack of technical skills, can you show a demo example, such as printing key and value when adding.
Or that this case is not suitable for using caffeine, it would be better to use map.put directly.
Thanks.

**Comment by ben-manes** [comment, marked as answer]
Hi @WeiMingzhi,

```java
Cache<String, Integer> cache = Caffeine.newBuilder().build();
cache.asMap().compute("a", (key, oldValue) -> {
  int newValue = 1;
  System.out.printf("%s: %d => %d%n", key, oldValue, newValue);
  return newValue;
});
cache.asMap().compute("a", (key, oldValue) -> {
  int newValue = 2;
  System.out.printf("%s: %d => %d%n", key, oldValue, newValue);
  return newValue;
});
System.out.printf("Cache contains: %s%n", cache.asMap());

cache.asMap().compute("a", (key, oldValue) -> {
  System.out.printf("%s: %d => %d%n", key, oldValue, null);
  return null;
});
System.out.printf("Cache contains: %s%n", cache.asMap());
```

```sh
a: null => 1
a: 1 => 2
Cache contains: {a=2}
a: 2 => null
Cache contains: {}
```

I hope that helps clarify things. Please let me know if you have further questions.

**Reply by WeiMingzhi** [reply]
Thanks, @ben-manes 
This usage still manipulates the Map every time and is not easy to use in a project.
However, the features I described are not required in use.
Thanks again for your reply!
